docs: what / why proposal for stream token counting via SSE eent parsing (#611)
shaneutt/praxis: docs/proposals/00211_streaming-token-counting.md

---
issue: https://github.com/praxis-proxy/praxis/issues/211
discussion: https://github.com/praxis-proxy/praxis/issues/20
status: proposed
authors:
  - szedan-rh
graduation_criteria:
  - SSE parser correctly accumulates token counts for all five providers
  - Both delta and cumulative provider semantics handled
  - Stream completion detection works for all sentinel formats
  - Token counts written to FilterContext via set_token_usage()
  - Non-streaming responses parsed in the same filter
  - Unit tests cover all provider streaming formats
stakeholders:

>   - shaneutt

Reviewer: ```suggestion
  - shaneutt
  - twghu
```

Author: Updated.

---

# Streaming Token Counting via SSE Event Parsing

## What?

A `token_count` filter that extracts token usage from AI provider

> response bodies and writes aggregated counts to `FilterContext`.

Reviewer: This is describing "How?" a bit too much for the "What?" section:

```suggestion
response bodies and stores aggregated counts for later consult.
```

Author: Updated.

This is the filter that connects the existing foundation pieces:
the provider parsing library ([#216], merged as [#510]) and the
`set_token_usage()` context method ([#212], merged as [#474]).

### Goals

- Parse SSE `data:` events for incremental token usage fields
  from all five supported providers (OpenAI, Anthropic, Google
  Gemini, Bedrock, Azure).
- Accumulate across chunks with correct semantics: override for
  cumulative providers, sum for delta providers.
- Emit final aggregated counts on stream completion (`[DONE]`
  sentinel, Anthropic `message_stop`, Bedrock `messageStop`,
  or stream close).
- Handle non-streaming JSON responses by parsing the complete
  response body and extracting usage in a single pass.
- Write final counts to `FilterContext` via `set_token_usage()`
  so downstream filters (rate limiting, logging, cost tracking)
  can consume them without provider coupling.
- Be transparent: response bodies and status codes pass through
  unchanged.

### Non-Goals

- Client-side token estimation before the response arrives
  ([#219]) -- separate filter, separate concern.
- Injecting `X-Token-*` response headers ([#214]) -- separate

>   filter that reads from `FilterContext`.

Reviewer: Similar to above, this non-goal speaks too directly to _how_ we would fulfill the goal. This needs to describe at a higher level what we want to do (or in this case, not do) and then later in the "How?" section we can decide how to resolve that goal.

Author: Updated.

- Token-based rate limiting ([#21]) -- downstream consumer of
  the counts this filter produces.
- Full response normalization across providers -- this filter
  extracts token usage only.

## Why?

### Motivation

**Streaming is the dominant path for AI inference.** In

> production, 80-90%+ of inference requests use `stream: true`.

Reviewer: Can you add a citation for this?
